### PR TITLE
gh-64376: Use Argument Clinic for datetime.date classmethods

### PR DESCRIPTION
**Pull request by hauntsaninja**
https://bugs.python.org/issue20177

* Issue: gh-64376

### CODE REVIEW

**Comment by hauntsaninja** [comment]
Re: `datetime.date.fromisocalendar`
Note that this will now throw OverflowError instead of ValueError when
using the C version of datetime. There is precedent for this, e.g.,
`datetime.date.fromordinal` will throw an OverflowError with the C module
but ValueError with the Python.

**Comment by hauntsaninja** [comment]
Two small notes:
1) Let me know if this diff size is too small and piece-meal-y!
2) I like the docstrings in Lib/datetime.py better, but unless otherwise told, I'll save that for another issue.

**Comment by taleinat** [comment]
> 2. I like the docstrings in Lib/datetime.py better, but unless otherwise told, I'll save that for another issue.

Indeed, it's good to keep these Argument Clinic conversions as straightforward as possible.

**Comment by taleinat** [comment]
Closing and re-opening to restart the hanged Travis-CI check.

**Comment by hauntsaninja** [comment]
Thanks, looks like everything's green :-)

File: Lib/test/datetimetester.py [hunk 1]
```diff
@@ -1909,9 +1909,6 @@ def test_fromisocalendar_value_errors(self):
             (10000, 1, 1),
             (0, 1, 1),
             (9999999, 1, 1),
-            (2<<32, 1, 1),
-            (2019, 2<<32, 1),
-            (2019, 1, 2<<32),
```
Comment (taleinat) [inline Lib/test/datetimetester.py:1914]: Why are these removed?

File: Misc/NEWS.d/next/Library/2020-05-19-06-32-11.bpo-20177.BwGrIX.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Convert :class:`datetime.date` class methods to use Argument Clinic.
```
Comment (taleinat) [inline Misc/NEWS.d/next/Library/2020-05-19-06-32-11.bpo-20177.BwGrIX.rst]: Our good bot Bedevere doesn't know this, but changes like these don't require a NEWS entry. I'll add the required GitHub, label, and this file should be removed.

File: Modules/_datetimemodule.c [hunk 3]
```diff
@@ -2876,13 +2876,18 @@ date_fromtimestamp(PyObject *cls, PyObject *obj)
...
-/* Return new date from current time.
- * We say this is equivalent to fromtimestamp(time.time()), and the
- * only way to be sure of that is to *call* time.time().  That's not
- * generally the same as calling C's time.
- */
+/*[clinic input]
+@classmethod
+datetime.date.today
+
+Return new date from current time.
```
Comment (taleinat) [inline Modules/_datetimemodule.c]: This can also be a datetime, since datetime subclasses date.

This PR otherwise retains the existing doc-strings, which I suggest doing consistently.

File: Modules/_datetimemodule.c [hunk 4]
```diff
@@ -2998,33 +3023,32 @@ date_fromisoformat(PyObject *cls, PyObject *dtstr)
...
+    year: int
+    week: int
+    day: int
 
-        }
-        return NULL;
-    }
+int, int, int -> Construct a date from the ISO year, week number and weekday.
+
+This is the inverse of the date.isocalendar() function
```
Comment (taleinat) [inline Modules/_datetimemodule.c]:
```suggestion
This is the inverse of the date.isocalendar() function.
```

**Review by taleinat — CHANGES_REQUESTED** [review]
Thanks for this!

Needs a few tweaks but overall looks good.

File: Lib/test/datetimetester.py (hunk 1 above)
Comment (hauntsaninja) [reply Lib/test/datetimetester.py:1914]: I mentioned this in a PR comment:

Re: `datetime.date.fromisocalendar`
Note that this now can throw OverflowError instead of ValueError when
using the C version of datetime. There is precedent for this, e.g.,
`datetime.date.fromordinal` will throw an OverflowError with the C module
but ValueError with the Python.
Also, prior to this PR, the C version `fromisocalendar` would throw a ValueError while the Python version throws TypeError for issues with e.g, missing arguments.

File: Modules/_datetimemodule.c [hunk 5]
```diff
@@ -2940,46 +2949,62 @@ datetime_date_fromtimestamp_capi(PyObject *cls, PyObject *args)
...
     }
     return result;
 }
 
-/* Return the new date from a string as generated by date.isoformat() */
+/*[clinic input]
+@classmethod
+datetime.date.fromisoformat
+
+    date_string: object
```
Comment (taleinat) [inline Modules/_datetimemodule.c]: We can use `date_string as dstr` here to avoid changing the C variable name.

```suggestion
    date_string as dstr: object
```

Comment (hauntsaninja) [reply Modules/_datetimemodule.c]: Is that preferable? datetime.py also uses `date_string` as the variable name in `fromisoformat`.

**Review by hauntsaninja — COMMENTED** [review]
Thanks for the review! :-)

File: Modules/_datetimemodule.c (hunk 5 above)
Comment (taleinat) [reply Modules/_datetimemodule.c]: Yes, it is preferable.

It won't be user-visible, it will only keep the internal C variable name `dstr`, to avoid unnecessary code churn.

Comment (hauntsaninja) [reply Modules/_datetimemodule.c]: Addressed!

File: Lib/test/datetimetester.py (hunk 1 above)
Comment (taleinat) [reply Lib/test/datetimetester.py:1914]: Ah, thanks, I didn't make the connection! I'll take another look at that.

**Comment by taleinat** [comment]
> 1. Let me know if this diff size is too small and piece-meal-y!

In my opinion this is fine, huge PRs can be hard to work with.

File: Modules/_datetimemodule.c [hunk 6]
```diff
@@ -2998,33 +3023,32 @@ date_fromisoformat(PyObject *cls, PyObject *dtstr)
...
-                keywords,
-                &year, &week, &day) == 0) {
-        if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
-            PyErr_Format(PyExc_ValueError,
-                    "ISO calendar component out of range");
+    year: int
+    week: int
+    day: int
 
-        }
```
Comment (taleinat) [inline Modules/_datetimemodule.c:3024]: Now I see that before this PR we have code in `fromisocalendar()` to raise ValueError rather than OverflowError. Raising a ValueError seems like better behavior to me. Also, changing this would be considered backwards-incompatible, would require NEWS and What's New entries, and couldn't be backported.

I think we should find a way to retain this behavior, even if that means not using Argument Clinic for this method right now.

**Comment by taleinat** [comment]
To respond to all points re. types of exceptions raised:

> Note that this now can throw OverflowError instead of ValueError when using the C version of datetime.

As I stated above, I suggest keeping the behavior of raising ValueError rather than OverflowError.

> There is precedent for this, e.g., datetime.date.fromordinal will throw an OverflowError with the C module but ValueError with the Python.

Changing this to raise ValueError instead of OverflowError would be an improvement IMO.

> Also, prior to this PR, the C version fromisocalendar would throw a ValueError while the Python version throws TypeError for issues with e.g, missing arguments.

These inconsistencies are unfortunate, but relatively minor. If this PR fixes that, adding a test for this would be great!
